### PR TITLE
`walrus-meter`

### LINKED ISSUE #4742: Applications don't load with Rust 1.87 or later
Starting from toolchain 1.87 upwards the resulting Wasm is not compatible with the network. It throws the following error when trying to create the instance of the contract:

```
2025-10-06T15:52:25.971095Z  INFO linera::main: linera_client::client_context: Module published successfully!
2025-10-06T15:52:25.971144Z  INFO linera::main: linera_client::client_context: Synchronizing client and processing inbox
2025-10-06T15:52:26.244197Z  INFO linera::main: linera_core::client: Received 0 certificates from 8 validator(s).
2025-10-06T15:52:26.614261Z ERROR linera: Error is Failed to create application

Caused by:
    chain client error: Local node operation failed: Worker operation failed: Execution error: Invalid Wasm module: Unknown opcode 252 during Operation(0)
```

You can see it publishes just fine and fails later – which is not ideal.To fix this, the rust toolchain used has to be 1.86 at most. Anything after that seems to produce a Wasm code that contains the disallowed "memory bulk" operation.

### PR DESCRIPTION
**Pull request by Twey**
## Motivation

We'd like upgrade our Rust version.  Currently we're being held back by `wasm-instrument`, which depends on `parity-wasm`, which is unmaintained and doesn't support some of the newer extensions to Wasm that the Rust compiler now emits by default.

## Proposal

Replace it with the new crate `walrus-meter`, which does the same thing but built on the maintained `walrus` library.

As a bonus, `walrus` intends to eventually support preserving DWARF debug info through its transformations.

## Test Plan

`walrus-meter` includes tests for compatibility with (how we use) `wasm-instrument`.  We have existing tests in this codebase, run in CI, that ensure that fuel costs don't change.  Those tests have in fact changed in this PR because `wasm-instrument` overcharges in some corner cases:

```
;; wasm-instrument          ;; walrus-meter
block  ;; label = @3        block  ;; label = @3
  i64.const $FOO_BAR_COST |   i64.const $FOO_COST
  call $spend                 call $spend
  block  ;; label = @4        block  ;; label = @4
    block  ;; label = @5        block  ;; label = @5
      ;; … FOO …                  ;; … FOO …
      br_if 0 (;@5;)              br_if 0 (;@5;)
      unreachable                 unreachable
    end                         end
    br 1 (;@3;)                 br 1 (;@3;)
  end                         end
                          >   i64.const $BAR_COST
                          >   call $spend
  ;; … BAR …                  ;; … BAR …
end                         end
```

This is a bug in the `wasm-instrument` case: the `br 1` will, if reached, exit `@3` causing `BAR` to be jumped over, but `wasm-instrument` has already merged the cost of it into the minimum cost of block `@3`, meaning `wasm-instrument` is charging for instructions that aren't executed.

In `linera-counter` the cost of `BAR` is 42, resulting in a 168 fuel discrepancy, i.e. 42×4 (the test does 4 iterations).

This may result in this PR being incompatible with Testnet Conway, if we assume that fuel costs don't change.

## Release Plan

- Nothing to do / These changes follow the usual release cycle.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)
- fixes https://github.com/linera-io/linera-protocol/issues/4742

### CODE REVIEW

File: linera-execution/src/wasm/mod.rs [hunk 1]
```diff
@@ -209,45 +208,14 @@ impl UserServiceModule for WasmServiceModule {
...
-        parity_wasm::deserialize_buffer(&bytecode.bytes)?,
-        gas_metering::host_function::Injector::new(
-            "linera:app/contract-runtime-api",
-            "consume-fuel",
-        ),
-        &WasmtimeRules,
+    let instrumented_module = walrus_meter::instrument(
+        &bytecode.bytes,
+        walrus_meter::costs::Wasmtime,
+        ("linera:app/contract-runtime-api", "consume-fuel"),
```
Comment (deuszx) [inline linera-execution/src/wasm/mod.rs:225]: So how are costs set for walrus? seems like the defaults are all zeros?

Comment (Twey) [reply linera-execution/src/wasm/mod.rs:225]: I built our ‘Wasmtime costs’ into `walrus-meter` as `walrus_meter::costs::Wasmtime`.

Mind you our Wasmtime costs don't actually represent Wasmtime because Wasmtime counts falling off the end of a function as a `return`, which we don't do (with `wasm-instrument` or `walrus-meter`).  With that in mind I might rename it.

Comment (deuszx) [reply linera-execution/src/wasm/mod.rs:225]: Does it mean that if we ever want to change the costs we need to modify the `walrus-meter` crate?

Comment (deuszx) [reply linera-execution/src/wasm/mod.rs:225]: Can we do it like it was with parity-wasm? I.e. have `walrus-meter` expose an interface and then implement it on our end?

Comment (Twey) [reply linera-execution/src/wasm/mod.rs:225]: That is how it works, it just comes with some default implementations.

Comment (deuszx) [reply linera-execution/src/wasm/mod.rs:225]: OK but why not make it clear what actual costs are? I think hiding it behind the default implementation is confusing. Especially that it's no longer how it used to be.

Comment (Twey) [reply linera-execution/src/wasm/mod.rs:225]: Sure, done. It adds a dependency on `walrus` but probably that's okay.

Comment (deuszx) [reply linera-execution/src/wasm/mod.rs:225]: Thank you.

**Review by ma2bd — APPROVED** [review]

**Review by deuszx — APPROVED** [review]
